Fix doc-gen errors and warnings

petebacondarwin opened the pull request:
See individual commits.

Fixes #39610 

Also see https://pr40404-d213226.ngbuilds.io/api/core/Sanitizer

petebacondarwin committed:
build(docs-infra): do not display stack trace on doc-gen error

This commit updates `dgeni` to the latest version, which
will now only display the stack-trace after an error if the
logger is set to `debug`.

Fixes #39610

petebacondarwin committed:
build(docs-infra): add `@default` tag-def support

This tag-def is being used in the `platform-server` source code,
and was causing warnings in the doc-gen. Adding support
prevents the warning from being shown.
Other than that this change has no effect on the generated
documentation, since the `@default` tag is not used in any
templates.

petebacondarwin committed:
build(docs-infra): ensure private exported subclasses are not rendered

Previously, only the first layer of sub-classes were filtered to remove
private exports. But on recursing down further levels these were not
filtered out.

mary-poppins commented:
You can preview d213226 at https://pr40404-d213226.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-api-package/tag-defs/default.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = function() {
+  return {name: 'default'};
```
Couldn't we also capture the default values and show them in the docs (not necessarily as part of this PR)?

petebacondarwin replied:
We could - in another PR :-D

gkalpak commented on aio/tools/transforms/templates/api/lib/descendants.html:
```diff
@@ -4,7 +4,7 @@
     <li>
       <code>{$ descendant.name $}</code>
-      {$ renderDescendantList(descendant.descendants | filterByPropertyValue('docType', docTypeMatcher), docType, recursed, docTypeMatcher) $}
+      {$ renderDescendantList(descendant.descendants | filterByPropertyValue('docType', docTypeMatcher) | filterByPropertyValue('privateExport', undefined), docType, recursed, docTypeMatcher) $}
```
Could we also make [this][1] a hard failure (instead of a warning) now? :pray:

[1]: https://github.com/angular/angular/blob/b48eabddb83900496e5e00fb98e192c6c49dff26/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js#L117-L118

petebacondarwin replied:
Adding

petebacondarwin replied:
Done - https://github.com/angular/angular/pull/40404/commits/939d58a31084b58fddbad57396efd5d3a058ed56

gkalpak reviewed (approved):
A couple of small suggestions. Otherwise lgtm :+1: :tada:
BTW, there is a typo in the second commit message: `has no affect` --> `has no effect`

petebacondarwin committed:
build(docs-infra): fail if we attempt to autolink to a private doc

Previously we only logged a warning if we attempted to auto-link
to a doc that had no `path` property, which implies that it is private
and is not rendered.

Now we fail hard during full doc-gen, although when running the
`yarn serv-and-sync` command it should not fail if changes are
only made to guides, which is what authors who use this tool
are most likely to do.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -19,12 +19,24 @@ const textContent = require('hast-util-to-string');
+ * 
+ * @property ignoredLanguages an array of languages that should not be auto-linked
+ * 
+ *
```
```suggestion
```

petebacondarwin replied:
Agh!!

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -19,12 +19,24 @@ const textContent = require('hast-util-to-string');
   autoLinkCodeImpl.customFilters = [];
   autoLinkCodeImpl.codeElements = ['code'];
   autoLinkCodeImpl.ignoredLanguages = ['bash', 'sh', 'shell', 'json', 'markdown'];
+  autoLinkCodeImpl.failOnMissingDocPath = false;
```
OOC, why not default to `true` (and override to `false` in the `authors-package`)?

petebacondarwin replied:
It turns out that changes to guides do not trigger auto-linking of code during `yarn serve-and-sync` so there was no need to change it for the author's package.

I did this like this so that we had a clear place in the base-package to turn it off if we needed to. Rather than having to search around inside the post-processors hoping for a switch.

gkalpak replied:
I generally prefer been stricter by default, but I don't feel strongly about this one. I'm fine keeping it as is :+1:

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -19,12 +19,24 @@ const textContent = require('hast-util-to-string');
  * Only text contained in these elements will be linked to.
  * Usually set to "code" but also "code-example" for angular.io.
+ * 
+ * @property ignoredLanguages an array of languages that should not be auto-linked
```
:100:

mary-poppins commented:
You can preview 544e4a6 at https://pr40404-544e4a6.ngbuilds.io/.

mary-poppins commented:
You can preview ee69b93 at https://pr40404-ee69b93.ngbuilds.io/.

mary-poppins commented:
You can preview 9321503 at https://pr40404-9321503.ngbuilds.io/.
